[blanklogic] iP

blanklogic opened the pull request:
# Friday
> "You can do anything, but not everything." – David Allen ([source](https://www.goodreads.com/quotes/242674-you-can-do-anything-but-not-everything))

Friday is designed to help you manage your tasks efficiently so that you can focus on what truly matters. It offers a **robust** and **intuitive** interface that is:

- Text-based, for quick and efficient input
- Easy to learn, with a simple command syntax
- ~FAST~ _SUPER_ FAST, enabling you to manage your tasks with minimal effort

## Getting Started

1. [Download the latest release](https://github.com/blanklogic/ip/releases)
2. Double-click the JAR file to launch the application
3. Add your tasks using simple commands
4. Let Friday handle the rest for you! :rocket:

AND most importantly, it is **FREE**!

## Key Features

- [x] **Task management**: Add, find, mark, unmark, and delete tasks effortlessly.
- [x] **GUI**: Graphical User Interface.
---
If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        new Friday().run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

blanklogic committed:
Level 0. Renamed Duke to Friday, Implemented initial greeting and exit skeletal message.

blanklogic committed:
Level 1. Implemented logic for accepting and echoing inputs of the user as well as exiting once exit flag has been triggered by the user

blanklogic committed:
Level 2. Implemented ability to log user input and displaying them back to the user when requested

blanklogic committed:
Level 3. Mark as Done. Implemented Task class and refactored code to utilize OOP concepts and modularized code for improved readability

blanklogic committed:
Level 4. Added support for tracking ToDos, Deadlines, and Events, refactored code for improved readability, and added comments to easily identify and prevent regression

blanklogic committed:
Updated TextUiTesting using I/O redirection technique with expected inputs and outputs from previous levels to ensure that code does not suffer from regression bugs

blanklogic committed:
Level 5. Implemented Error Handling, Phased out legacy Level 2 for Level 4 implementation of Task list, Added 'help' for list of commands.

blanklogic committed:
Level 6. Implemented Delete functionality with appropriate error handling, included in list of commands

blanklogic committed:
Refactor: Utilise enums for command handling

blanklogic committed:
Update: Implement auto-save and load functionality for tasks

The application does not currently persist tasks between sessions, leading
to a loss of data when the application is closed.

Persisting tasks to the hard disk ensures that the user's data is saved
automatically whenever the task list changes, enhancing the user
experience.

This commit introduces the following changes:
* Implement saving tasks to the hard disk whenever the task list changes.
* Load tasks from the hard disk when the chatbot starts up.
* Hardcode the file path as ./data/friday.txt.
* Update .gitignore to exclude the constantly updating ./data/friday.txt.

The file path is hardcoded to simplify the implementation, and the
.gitignore update prevents unnecessary tracking of the data file.

blanklogic committed:
Merge branch 'branch-Level-7'

blanklogic committed:
Update: Add support for parsing and handling dates and times

blanklogic committed:
Merge branch 'branch-Level-8'

blanklogic committed:
Refactor codebase to improve OOP by extracting related code into classes

blanklogic committed:
Merge branch 'branch-A-MoreOOP'

blanklogic committed:
Organize classes into packages for better modularity and structure

blanklogic committed:
Merge branch 'branch-A-Packages'

blanklogic committed:
Merge branch 'add-gradle-support'

blanklogic committed:
Add JUnit tests for all non-trivial methods of all classes

blanklogic committed:
Pack application as executable JAR file

blanklogic committed:
Add JavaDoc comments to all public classes and methods

blanklogic committed:
Enhance code to comply with the SE-EDU Java coding standard

blanklogic committed:
Implement find functionality for tasks

blanklogic committed:
Merge branch 'branch-A-JavaDoc'

blanklogic committed:
Merge branch 'branch-A-CodingStandard'

blanklogic committed:
Merge branch 'branch-Level-9'

blanklogic committed:
Fix storage saving issues and missing packages

blanklogic committed:
Refactor Parser to remove wildcard import

blanklogic committed:
Refactor JUnit Tests to remove wildcard imports

blanklogic committed:
Apply CheckStyle fixes and add JavaDoc to FindCommand.java

blanklogic committed:
Merge branch 'branch-A-CheckStyle'

blanklogic committed:
Implement an initial GUI for Friday chat bot assistant

blanklogic committed:
Merge branch 'branch-Level-10'

SherniceSng commented on src/main/java/friday/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,56 @@
+        if (inputs.length != 2) {
+            throw new FridayException("\tInvalid input. 'deadline' command requires a"
+                    + " description and deadline\n\tusage: deadline <string> /by <yyyy-MM-dd HHmm>.");
+        }
```
I like how the indentation here follows the coding standard.

SherniceSng commented on src/main/java/friday/command/ByeCommand.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    public boolean isExit() {
+        return true;
+    }
```
The naming of boolean is good.

SherniceSng commented on src/main/java/friday/Friday.java:
```diff
@@ -0,0 +1,60 @@
+ */
+public class Friday {
+    private final Storage storage;
+    private TaskList tasks;
```
The naming of a collection of objects using plural is good.

SherniceSng commented on src/main/java/friday/task/TaskList.java:
```diff
@@ -0,0 +1,105 @@
+ * Represents a list of tasks.
+ */
+public class TaskList {
+    private final ArrayList<Task> tasks;
```
The use plural for a collection of objects is good here too.

SherniceSng commented on src/main/java/friday/command/Parser.java:
```diff
@@ -0,0 +1,83 @@
+        default ->
+                throw new FridayException("Invalid input. Please ensure that this command is supported by me"
+                        + " and you have utilized the right syntax.\nCheck 'help' for more information.");
+        };
```
Should the switch block be changed slightly to better fit the coding standard?

SherniceSng reviewed (commented):
LGTM. Keep up the good work!

blanklogic committed:
Fix UiTest.java to fit the new requirements of a GUI

blanklogic committed:
Utilise assert for documenting important assumptions

Add assert statements in key parts of the code to document and
enforce important assumptions. This includes:

- Ensuring input values meet expected conditions.
- Validating preconditions before executing logic.

Assertions are used to catch unexpected conditions during
development and testing, helping to identify issues early.

This change improves code reliability by making assumptions explicit,
which aids in debugging and maintaining the codebase.

blanklogic committed:
Refactor code to adhere to SLAP and improve readability

Refactor methods to comply with the Single Level of Abstraction Per
Method (SLAP) principle.

This involves:

- Reducing method nesting and simplifying complex logic.
- Breaking long methods into smaller, more focused methods.
- Applying consistent formatting and naming conventions.

The refactoring enhances code readability and maintainability by
reducing complexity and making each method easier to understand.

These changes also align the code with SE-Education Java Coding
Standards, making future modifications and reviews more straightforward.

blanklogic committed:
Merge pull request #1 from blanklogic/branch-A-Assertions

Utilise assert for documenting important assumptions

blanklogic committed:
Merge branch 'master' into branch-A-CodeQuality

blanklogic committed:
Merge pull request #2 from blanklogic/branch-A-CodeQuality

Refactor code to adhere to SLAP and improve readability

blanklogic committed:
Set up Continuous Integration (CI) using GitHub Actions

blanklogic committed:
Fix Code styling issues to adhere to SE-EDU Coding Standards

blanklogic committed:
Resolve CI issues on GitHub Actions by including friday.txt

blanklogic committed:
Resolve JUnit Test issues due to regression

blanklogic committed:
Implement C-Update to enable editing of existing tasks

The current system does not support the modification of existing tasks,
forcing users to delete and recreate tasks for any changes.
This leads to unnecessary duplication and inconvenience.

Let's introduce the UpdateCommand, which allows users to easily modify
the details of a task (such as description, deadline, or event time)
without needing to delete and recreate it.

- Added the UpdateCommand class to handle the update operation.
- Users can update a task by specifying the task number followed by
new details.
- Modified the system to handle longer descriptions and multiple task
types (Todo, Deadline, Event).
- Tasks are saved to storage after successful updates.

This new feature improves the flexibility of Friday, the task manager,
enhancing the user experience by allowing seamless task updates.

blanklogic committed:
Merge branch 'branch-C-Update'

blanklogic committed:
Add a Friday Task Manager User Guide

blanklogic committed:
Update User Guide with Quick Start

blanklogic committed:
Fix Typo in User Guide
